[5.0] Performance optimizations (14% for hello world L5 app)

neomerx opened the pull request:
Optimizations for improve performance for 14.3%

Detailed description https://github.com/neomerx/rhw-l5#researching-laravel-5-framework

neomerx committed:
Improve performance of default Laravel application for 4.3%

For test details see https://github.com/neomerx/rhw-l5#illuminatecontainercontainerfireresolvingcallbacks

neomerx committed:
Improve performance of default Laravel application for 1.4%

For test details see https://github.com/neomerx/rhw-l5#illuminateroutingroutergatherroutemiddlewares

neomerx committed:
Remove unnecessary array creation which is created recursively on every request.

For test details see https://github.com/neomerx/rhw-l5#illuminatepipelinepipelinethen

neomerx committed:
Improve performance of default Laravel application for 3.4%

For test details see https://github.com/neomerx/rhw-l5#illuminatefoundationapplicationgetprovider

GrahamCampbell commented on src/Illuminate/Pipeline/Pipeline.php:
```diff
@@ -92,13 +92,13 @@ public function via($method)
 	{
-		$firstSlice = $this->getInitialSlice($destination);
+		$linkNext = $this->getSlice();
+		$next     = $this->getInitialSlice($destination);
```
do not align these please

GrahamCampbell commented on src/Illuminate/Pipeline/Pipeline.php:
```diff
@@ -92,13 +92,13 @@ public function via($method)
-		$firstSlice = $this->getInitialSlice($destination);
+		$linkNext = $this->getSlice();
+		$next     = $this->getInitialSlice($destination);
+		for ($i = count($this->pipes) - 1; $i >= 0; $i--) {
```
Using preincrementation would be never faster technically.

neomerx replied:
Main gain here from dropping multiple array creation on every request.
See https://github.com/neomerx/rhw-l5#illuminatepipelinepipelinethen

GrahamCampbell replied:
Yeh, absolutely.

GrahamCampbell commented:
This looks really nice. Thank you!

KennedyTedesco commented:
Really impressive, @neomerx :beer:

EliuFlorez commented:
+1

taylorotwell commented:
Thank you very much for taking the time to look into this. Will look it over in more detail tonight.

djtarazona commented:
:+1:

overtrue commented:
:100:

taiyaei commented:
Wow, very nice. If those numbers are consistent then that is very impressive!

DearMadMan commented:
really nice

phanan commented on src/Illuminate/Routing/Router.php:
```diff
@@ -702,11 +701,14 @@ protected function runRouteWithinStack(Route $route, Request $request)
+		foreach ($route->middleware() as $tag)
 		{
-			return Collection::make(array_get($this->middleware, $m, $m));
+			$result[] = (isset($this->middleware[$tag]) === true ? $this->middleware[$tag] : $tag);
```
Do we need to explicitly compare to `true`, knowing that `isset()` returns `BOOL`?

phanan commented:
:+1: Always a fan of optimization.

CurosMJ commented:
+1 Appreciate that you took time out to do this.

garygreen commented:
I'm a definitely +1 to all performance optimizations but find the title really misleading. It gives the impression your Laravel application will be +14% faster which is not the case in real world. We're talking about a total of 5ms difference between all the commits, something probably completely inconceivable in real world usage.

Arrilot commented:
Yeah, Gary is right, it's not even fresh L5 install.
I bet that Eloquent is the slowest part of the framework in an average app.
Still impressive

neomerx commented:
@garygreen Take a look at https://github.com/kenjis/php-framework-benchmark#hello-world-benchmark
What you call 'inconceivable' milliseconds is actually difference between 'good' and 'terribly slow' for php public opinion

garygreen commented:
@neomerx So how many requests per second does Laravel go up by in that chart, with these commits?

garygreen commented:
Just to clarify, I have zero issue with this PR at all, welcome any performance improvements. The only thing is the commit names/title of this PR being misleading to others, looking through the commit history in Laravel. "Improve performance of default Laravel application for 4.3%" Those are pretty sweeping statements that are very misleading out of context when you look at the numbers.

neomerx commented:
@garygreen feel free to make your pull request for contributors' naming guidelines which not invulnerable for critics by itself

JoostK commented:
This reintroduces the buggy #8135, see also #8210.

neomerx commented:
@JoostK I've downloaded fresh L5 app, applied the patch and before and after I got this
```
neomerx@L5:~/bug-fix/laravel$ php artisan event:generate
Events and handlers generated successfully!
```
Where is the bug?

neomerx commented:
@JoostK my patch and  #8135 have different implementations. Please either support your statement that    #8239 is buggy or remove it.

GrahamCampbell commented:
> So how many requests per second does Laravel go up by in that chart, with these commits?

That list is totally wrong anyway. Laravel does like 500 if you turn on route caching and config caching, and you strip out the middelwares.

neomerx commented:
@GrahamCampbell Moreover I have a feeling that due to their kernel settings you will have different result if just change test order in https://github.com/kenjis/php-framework-benchmark/blob/master/benchmarks/hello_world.sh

neomerx committed:
Fix alignment as requested by @GrahamCampbell

shehi commented:
+1

toddbc commented:
@neomerx please see the test added in #8210, which will show why abcf531890a31be32edc81a4ba2e45fcf2d46897 won't work.  I also fell into the same trap.

garygreen commented:
@GrahamCampbell out of interest how do you get 500 requests per second with Laravel? I can't get close to that amount, even with base L5 install and route, config, compiled optimisations?  I'm getting 40 rqs p/s.

GrahamCampbell commented:
@garygreen Disable all the middlewares, use apply all the optimisations, and change opcache to cache large files too. That should make a difference because the compiled.php file won't get cached by default because it's too large.

garygreen commented:
Ohh thanks! That's interesting to know about opcode cache. When you say disable all middlewares, isn't that a critical part / flow in your app or am I misunderstanding what you mean by that?

GrahamCampbell commented:
> When you say disable all middlewares, isn't that a critical part / flow in your app or am I misunderstanding what you mean by that?

Well, it makes a more fair comparison between the other frameworks.

garygreen commented:
Interesting, opcache by default [caches all files regardless of size](http://php.net/manual/en/opcache.configuration.php) `opcache.max_file_size`. Still even with middleswares disabled and opcache I can't get anywhere close to 500 rq/s. You must have some magical setup or something ;D

neomerx committed:
Add search provider by parent class or interface to getProvider

neomerx committed:
Add test searching provider by parent class or interface

franzliedke commented:
Also, why should we disable all middlewares?

neomerx commented:
@toddbc Thanks for good point that I missed support for searching by interface supported by provider.

It looks it is not used by the framework itself and I didn't have any issues neither with tests nor with ```event:generate```. However for backward compatibility I've added such support and it doesn't effect performance.

I've also added slightly modified version of your test to a separate commit and leave it up to @taylorotwell which test version to commit. Anyway copyright is yours )

neomerx commented:
@GrahamCampbell @garygreen From my experience the environment has a huge impact on performance behavior
* on real hardware I had better performance improvement (in %) than on virtualized. Possibly virtual network, memory and etc has its own limitations
* kernel settings have big impact on performance
* php settings has also significant effect on performance

From default LAMP L5 Hello World 75 RPS to 500 RPS on the same hardware? Ez. Drop unnecessary middleware, apply all L5 buit-in optimization, NGINX, HHVM, kernel setup. What if PHP but not HHVM? Well, that wouldn't easy but looks achievable on non-virtual infrastructure and careful PHP settings.

garygreen commented:
Maybe Graham was talking about HHVM. I'm testing on standard PHP. What PHP settings do you recommend to tweak specifically?

neomerx commented:
@garygreen I've did such testing some time ago and changed settings in ```/etc/php5/fpm/pool.d/www.conf```.  Sadly I've lost that settings but as I remember it's only a few parameters that start with 'pm.'

taylorotwell commented:
It's going to be dependent on your machine. I get 400-500 with everything
optimized and sessions disabled.

On Saturday, April 4, 2015, Gary Green <notifications@github.com> wrote:

> @GrahamCampbell <https://github.com/GrahamCampbell> out of interest how
> do you get 500 requests per second with Laravel? I can't get close to that
> amount, even with base L5 install and route, config, compiled
> optimisations? I'm getting 40 rqs p/s.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laravel/framework/pull/8239#issuecomment-89585555>.
>

toddbc commented:
@neomerx, I took a look at 29d5f863a19add2160a0365b3da8363d82ab11fc.  I'll just note that the common case for getProvider is looking up a provider that doesn't exist yet (on initial registration.)  It's barely called otherwise iirc.  This means it will always hit the foreach, since it will ultimately return `null`.

So, with `n` providers, it would end up doing `(n * (n - 1)) / 2` instanceof comparisons.  There's probably some breaking point where using `class_parents($class) + class_implements($class)` only `n` times will be faster, though would have to benchmark to see it.  Did a quick comparison since I was interested, suggests the breakoff might be around ~9 providers.  But I didn't test carefully, just saw that `$x = $y instanceof z` takes about 23% as long as `$x = class_parents('y') + class_implements('y') + array('y')`.

Strange that you didn't have problems with event:generate, though.  You're sure you're testing with a clean setup?

This line creates the provider:
https://github.com/laravel/framework/blob/614d08023f6e87cd535f38f875ba06c24e9c842f/src/Illuminate/Foundation/Console/EventGenerateCommand.php#L28

But normally this is an app-specific subclass:
https://github.com/laravel/laravel/blob/6449472ef7c7899785fb05bbed8eb7f5c3a08395/config/app.php#L145

I'm sure you might've nixed that unnecessary subclass for performance if you're not using events in that way, but in a default setup you would experience the problem.

neomerx committed:
Revert getProvider. This code for default application always return null and rather needs refactoring to eliminate it than optimization.

neomerx commented:
@toddbc Yep, it's always returns ```null```. For default application it's invoked 21 times and I think for typical application it would be somewhere in between 20-25 providers. It shouldn't work like this IMHO. Thus I tend to think it's better to refactor it and ask architect @taylorotwell for it/advise.

So it looks we don't have any open questions with this PR anymore and it could be merged, right?

Here is the list of all invoked providers

>getProvider Illuminate\Events\EventServiceProvider hit null
getProvider Illuminate\Routing\RoutingServiceProvider hit null
getProvider Illuminate\Auth\AuthServiceProvider hit null
getProvider Illuminate\Routing\ControllerServiceProvider hit null
getProvider Illuminate\Cookie\CookieServiceProvider hit null
getProvider Illuminate\Database\DatabaseServiceProvider hit null
getProvider Illuminate\Encryption\EncryptionServiceProvider hit null
getProvider Illuminate\Filesystem\FilesystemServiceProvider hit null
getProvider Illuminate\Foundation\Providers\FoundationServiceProvider hit null
getProvider Illuminate\Foundation\Providers\FormRequestServiceProvider hit null
getProvider Illuminate\Pagination\PaginationServiceProvider hit null
getProvider Illuminate\Session\SessionServiceProvider hit null
getProvider Illuminate\Validation\ValidationServiceProvider hit null
getProvider Illuminate\View\ViewServiceProvider hit null
getProvider App\Providers\AppServiceProvider hit null
getProvider App\Providers\BusServiceProvider hit null
getProvider App\Providers\ConfigServiceProvider hit null
getProvider App\Providers\EventServiceProvider hit null
getProvider App\Providers\RouteServiceProvider hit null
getProvider Illuminate\Bus\BusServiceProvider hit null
getProvider Illuminate\Hashing\HashServiceProvider hit null

rossedman commented:
:+1: Sweet.

GrahamCampbell commented:
This needs rebasing and squashing to one commit please.

GrahamCampbell commented:
I just tried this out and I don't see a 14% change. Using the config/route caching I see a change of only 3.7%.

neomerx commented:
@GrahamCampbell The methodology, used performance tools and the performance test results in the description. If you have found any errors there please be specific.

garygreen commented:
@GrahamCampbell as I've been mentioning... the title is misleading. If you look at the figures, it's not 14% at all. On average it's 4% and with that 4% is just a few milliseconds difference ;-)

toddbc commented:
I think it's a shame to close this.  Whether the optimizations amount to 4% of 14%, that's still an improvement.  So long as it doesn't reduce code quality free speed is always nice.

I'll say again that tests can benefit from a lot of these optimizations most, and I hate wasting time on tests.  Not only can tests not benefit from cached routes and etc. (well, I made route/config cache filenames include the environment name to improve this in my own build), but often they are testing small things, so the overall setup is accentuated.

More requests per second is nice, but even if that's small, I'll take a 4% reduction in test time any day.  I'm more than impatient enough to be happy to see my code works properly 2 seconds earlier.

Just IMHO.

onbjerg commented:
I agree with @toddbc

GrahamCampbell commented:
Why is this closed?

GrahamCampbell commented:
It was a certainly worth investigating. We're very grateful for the time you spent on this.

GrahamCampbell commented:
Are you planning to resubmit? I was just a little confused by the performance claims.

shehi commented:
I agree with everyone who says it needs further investigation and shouldn't be closed. The sheer attempt by the author to do such research shows how valuable the overall work and time put into it, are. I would love to see any percentage of performance: 1% translating to meer milliseconds in small setup can mean the same 1% in dozens, if not hundreds of those milliseconds. We should start somewhere, today with this 4% boost, tomorrow another 4% and at the end of the day we will have much better performing framework.

EliuFlorez commented:
+1 open

rossedman commented:
@shehi Agreed. The author closed it though. Hopefully @neomerx is resubmitting because it needed some rebasing and stuff like that.

acasar commented:
I really like the work @neomerx has done here. Hope this gets resubmitted.
